### PR TITLE
Allow to run e2e tests locally

### LINKED ISSUE #55511: testing: Run E2E tests locally
To enable teammates to write more E2E tests it needs to be easier to run the tests locally. At present, the E2E tests can only really run in CI since that is where all the credentials are needed to communicate with all the external dependencies for the test.

#### Open questions
- Retrieval of credentails
- Should the credentials be fetched from the devx-service
- Licensing issues due to limited number of seats
- For local testing, should be use a request replayer/mock server?

Will something like #55509 answer these open questions

### PR DESCRIPTION
**Pull request by jhchabran**
Follow-up to #55969
Fixes #55511

This PR makes it possible to run the e2e tests locally, assuming you have the proper secrets in your environment (this is to be fixed in another PR). 

- [x] `//testing:backend_integration_test` 
  - env needed: `GHE_GITHUB_TOKEN`, `SOURCEGRAPH_LICENSE_KEY`, `SOURCEGRAPH_LICENSE_GENERATION_KEY` 
- [x] `//testing:grpc_backend_integration_test`
  - same env as above 
- [x] `//testing:e2e_test` 
  - env needed: `GHE_GITHUB_TOKEN`, `SOURCEGRAPH_LICENSE_KEY`, `SOURCEGRAPH_LICENSE_GENERATION_KEY` + `GH_TOKEN`, `E2E_HEADLESS=bool` `E2E_SOURCEGRAPH_BASE_URL=http://localhost:...` 
- [x] `//testing:codeintel_integration_test`
  - env needed: same as backend integration + `GITHUB_TOKEN` + `GOOGLE_APPLICATION_CREDENTIALS` (we use the one from the local machine)  

It works by adding a new build setting that we can use to `select` the right `goarch/goos` on the `go_test` rules defining the test runners for those e2e tests. 

Sadly, the codeintel-qa one is a bit peculiar and requires to disable the sandbox locally, because you can't have a `select` in tags. It's fairly possible that we find a better approach down the line, see https://sourcegraph.slack.com/archives/CHXHX7XAS/p1692609624328779 but for the time being, it'll be fine. 

## Test plan

CI must pass (and I made it run the code intel QA on this branch to ensure it's not broken) 

CleanShot 2023-08-21 at 15 34 50@2x

## `//testing:backend_integration_test` 

```
~/work/sourcegraph U jh/local-e2e-runner $ bazel test //testing:backend_integration_test --config darwin-docker
INFO: Analyzed target //testing:backend_integration_test (0 packages loaded, 0 targets configured).
INFO: Found 1 test target...
Target //testing:backend_integration_test up-to-date:
  bazel-bin/testing/backend_integration_test
INFO: Elapsed time: 278.522s, Critical Path: 277.81s
INFO: 2 processes: 1 internal, 1 darwin-sandbox.
//testing:backend_integration_test                                       PASSED in 277.8s
  WARNING: //testing:backend_integration_test: Test execution time (277.8s excluding execution overhead) outside of range for MODERATE tests. Consider setting timeout="long" or size="large".

Executed 1 out of 1 test: 1 test passes.
INFO: Build completed successfully, 2 total actions
```

## `//testing:gprc_backend_integration_test`

```
~/work/sourcegraph U jh/local-e2e-runner $ bazel test //testing:grpc_backend_integration_test --config darwin-docker
INFO: Analyzed target //testing:grpc_backend_integration_test (4 packages loaded, 46215 targets configured).
INFO: Found 1 test target...
Target //testing:grpc_backend_integration_test up-to-date:
  bazel-bin/testing/grpc_backend_integration_test
INFO: Elapsed time: 328.412s, Critical Path: 310.00s
INFO: 5 processes: 4 internal, 1 darwin-sandbox.
//testing:grpc_backend_integration_test                                  PASSED in 286.6s
  WARNING: //testing:grpc_backend_integration_test: Test execution time (286.6s excluding execution overhead) outside of range for MODERATE tests. Consider setting timeout="long" or size="large".

Executed 1 out of 1 test: 1 test passes.
INFO: Build completed successfully, 5 total actions
```

## `//testing:e2e_test`

```
~/work/sourcegraph U jh/local-e2e-runner $ bazel test //testing:e2e_test --config darwin-docker --define=E2E_HEADLESS=false --define=E2E_SOURCEGRAPH_BASE_URL="http://localhost:7080" --define=GH_TOKEN=$GH_TOKEN
INFO: Analyzed target //testing:e2e_test (1 packages loaded, 4815 targets configured).
INFO: Found 1 test target...
Target //testing:e2e_test up-to-date:
  bazel-bin/testing/e2e_test
INFO: Elapsed time: 218.797s, Critical Path: 209.40s
INFO: 5 processes: 3 internal, 2 darwin-sandbox.
//testing:e2e_test                                                       PASSED in 170.8s

Executed 1 out of 1 test: 1 test passes.
INFO: Build completed successfully, 5 total actions
```

## `//testing:codeintel_integration_test_local` 

```
~/work/sourcegraph U jh/local-e2e-runner $ bazel test //testing:codeintel_integration_test_local --config darwin-docker                   
INFO: Analyzed target //testing:codeintel_integration_test_local (0 packages loaded, 0 targets configured).
INFO: Found 1 test target...
Target //testing:codeintel_integration_test_local up-to-date:
  bazel-bin/testing/codeintel_integration_test_local
INFO: Elapsed time: 222.448s, Critical Path: 221.74s
INFO: 2 processes: 2 darwin-sandbox.
//testing:codeintel_integration_test_local                               PASSED in 221.5s

Executed 1 out of 1 test: 1 test passes.
INFO: Build completed successfully, 2 total actions
```

### CODE REVIEW

**Comment by jhchabran** [comment]
@burmudar this PR is now ready, I just need to do a quick dive in the parent PR to see if we can avoid patching the hermetic-cc rules for local dev (the ones that disables ubsan)

**Comment by jhchabran** [comment]
@burmudar update: sadly, I can't go around patching hermetic_cc, not without spending more time on this, which for the moment I don't have. So this PR is actually ready :) We can merge it once reviewed and the parent is merged.

**Comment by jhchabran** [comment]
@mucles these targets all require env vars to be set, which are secrets we need to get injected in the local env. We can most probably fix that by wrapping these targets in `sg test` and handling the secrets fetching through authenticated calls to the devx-service.

**Comment by burmudar** [comment]
> @burmudar this PR is now ready, I just need to do a quick dive in the parent PR to see if we can avoid patching the hermetic-cc rules for local dev (the ones that disables ubsan)

Thanks so much @jhchabran! Just at the right time too! We can get this merged since I am now at the stage where I need a Sourcegraph running 🚀

File: BUILD.bazel [hunk 1]
```diff
@@ -286,6 +286,24 @@ go_proto_compiler(
     ],
 )
 
+# darwin_docker config
+#
+# To be able to run e2e tests locally, we need to simultaneously be able to build linux/amd64 container images
+# and to build go test runners targeting the local machine architecture and OS. Unless you're working on a linux machine
+# these two won't be the same. So we use that flag to indicate those particular runner to target darwin/arm64 if we see
```
Comment (burmudar) [inline BUILD.bazel:293]:
```suggestion
# these two won't be the same. So we use this flag to indicate that this particular runner should target darwin/arm64 if we see
```

File: BUILD.bazel [hunk 2]
```diff
@@ -286,6 +286,24 @@ go_proto_compiler(
...
+# have this build setting enabled.
```
Comment (burmudar) [inline BUILD.bazel:294]:
```suggestion
# that we have this build setting enabled.
```

File: dev/gqltest/BUILD.bazel [hunk 3]
```diff
@@ -37,3 +45,12 @@ go_test(
         "@io_k8s_utils//strings/slices",
     ],
 )
+
+genrule(
```
Comment (burmudar) [inline dev/gqltest/BUILD.bazel:49]: left over debugging?

File: testing/BUILD.bazel [hunk 4]
```diff
@@ -160,3 +166,89 @@ server_integration_test(
         "manual",
     ],
 )
+
+# This one is bit peculiar, when running the codeintel-qa locally, we need to disable the sandbox
```
Comment (burmudar) [inline testing/BUILD.bazel:170]:
```suggestion
# This one is bit perculiar, when running the codeintel-qa locally, we need to disable the sandbox
```

File: dev/gqltest/BUILD.bazel (hunk 3 above)
Comment (jhchabran) [reply dev/gqltest/BUILD.bazel:49]: Yes, thanks

**Review by burmudar — APPROVED** [review]
<3

**Comment by jhchabran** [comment]
@Strum355 I had to update the snapshots again, can you double check that it's correct?

File: docker-images/syntax-highlighter/crates/sg-syntax/src/snapshots/sg_syntax__sg_treesitter__test__rust.rs.snap
Comment (Strum355) [inline docker-images/syntax-highlighter/crates/sg-syntax/src/snapshots/sg_syntax__sg_treesitter__test__rust.rs.snap:1]: @tjdevries might need to update the queries for this 🤔 not sure what `assertion_line` is about

**Pull request merged**